File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge47.js
/*
 * Reto #47
 * VOCAL MÁS COMÚN
 * Fecha publicación enunciado: 21/11/22
 * Fecha publicación resolución: 28/11/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Crea un función que reciba un texto y retorne la vocal que más veces se repita.
 * - Ten cuidado con algunos casos especiales.
 * - Si no hay vocales podrá devolver vacío.
 *
 * Información adicional:
 * - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal"
 *   para preguntas, dudas o prestar ayuda a la comunidad.
 * - Tienes toda la información sobre los retos semanales en
 *   https://retosdeprogramacion.com/semanales2022.
 *
 */

function mostUsedVowel(text){
    let vowelCount = {}
    text.toLowerCase().normalize('NFD').split("").forEach(char => {
        if("aeiou".includes(char)){
            (char in vowelCount)? vowelCount[`${char}`]++ : vowelCount[`${char}`] = 1;
        }
    })
    let mostRepeated = [];
    let maxRepeated = 0;
    for(vowel in vowelCount){
        let count = vowelCount[vowel];
        if(count >= maxRepeated){
            if(count > maxRepeated) mostRepeated = [];
            mostRepeated.push(vowel);
            maxRepeated = count;
        }
    }
    return (mostRepeated.length !== 0)? mostRepeated : "There are no vowels";
}

console.log(mostUsedVowel("Hello")); // ['e', 'o']
console.log(mostUsedVowel("ghdjf")); // "There are no vowels"
console.log(mostUsedVowel("Hello World")); // ['o']
console.log(mostUsedVowel("This is a good day")); // ['i',  'a', 'o']
console.log(mostUsedVowel("Hola, ¿qué tal estás hoy?")); // ['a']

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge47.kt
package com.mouredev.weeklychallenge2022


/*
 * Reto #47
 * VOCAL MÁS COMÚN
 * Fecha publicación enunciado: 21/11/22
 * Fecha publicación resolución: 28/11/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Crea un función que reciba un texto y retorne la vocal que más veces se repita.
 * - Ten cuidado con algunos casos especiales.
 * - Si no hay vocales podrá devolver vacío.
 *
 * Información adicional:
 * - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal"
 *   para preguntas, dudas o prestar ayuda a la comunidad.
 * - Tienes toda la información sobre los retos semanales en
 *   https://retosdeprogramacion.com/semanales2022.
 *
 */

fun main(){
    mostRepeatedVocalIn("Enunciado: Crea un función que reciba un texto y retorne la vocal que más veces se repita." +
            " * Si no hay vocales podrá devolver vacío.")
    mostRepeatedVocalIn("")
}

private fun mostRepeatedVocalIn(textForAnalysis : String) : Char?{
    val totalLettersCharMap = mutableMapOf<Char, Int>()

    fun addOrCount(letter : Char){
        if (totalLettersCharMap.containsKey(letter)){
            totalLettersCharMap[letter] = totalLettersCharMap[letter]!!.plus(1)
        } else {
            totalLettersCharMap[letter] = 1
        }
    }

    textForAnalysis.split(" ").forEach {
        it.toCharArray().forEach { singleLetter ->
            addOrCount(singleLetter)
        }
    }

    var mostRepeated : Char? = null
    var totalCount = 0

    fun computeMostRepeatedVocal(){
        val vocals = listOf('a', 'e', 'i', 'o', 'u')
        vocals.forEach{ vocal ->
            if (totalLettersCharMap.containsKey(vocal)){
                if (totalLettersCharMap[vocal]!! > totalCount){
                    mostRepeated = vocal
                    totalCount = totalLettersCharMap[vocal]!!

                    println("$vocal , $totalCount")
                }
            }
        }
    }

    fun computeMostRepeatedLetter(){
        totalLettersCharMap.forEach{
            if (it.value > totalCount){
                totalCount = it.value
                mostRepeated = it.key
            }
        }
    }

    computeMostRepeatedVocal()
    if (mostRepeated != null)
        println("Int the text : \"$textForAnalysis\", " +
            "the most repeated vocal is $mostRepeated $totalCount times")
    else
        println("Null")

    return mostRepeated
}